### PR TITLE
910 Created menstrual hygiene products calculator

### LINKED ISSUE #910: [MHC] Сalculations and their results
As a **user** I want to **see the calculation results** so that **I will make my conclusions**

**Description**
1. User enters data
2. Program calculates the results using data
3. When user **clicks on the "start" button** program **scrolls down** and **shows the results** 

**Acceptance criteria**
After entering all the data, the calculator displays 4 indicators: 
1. how many disposable menstrual products have been used since the onset of menstruation until the current age
2. how much money has already been spent on disposable menstrual products from the time of menstruation to the current age
3. how much more menstrual products can be used from the current age to the expected age of menopause
4. how much money can still be spent on menstrual products from the current age to the expected age of menopause

### How to calculate:

1. **Загальна кількість менструацій від початку до менопаузи**
(Очікуваний вік менопаузи−Вік настання менструації)∗(365/Середня тривалість менструального циклу)
2. **Загальна кількість одноразових продуктів**
(Кількість одноразових продуктів за цикл∗Загальна кількість менструацій)
3. **Загальна вартість одноразових продуктів**
(Загальна кількість одноразових продуктів∗Вартість одного одноразового продукту )
4. **Загальна кількість багаторазових продуктів**
(Очікуваний вік менопаузи−Вік настання менструації)/ Термін служби одного багаторазового продукту 
5. **Загальна вартість багаторазових продуктів:** 
(Загальна кількість багаторазових продуктів* Загальна вартість одноразових продуктів)
6. **Фінансова вигода** 
(Загальна вартість одноразових продуктів- Загальна вартість багаторазових продуктів)

### PR DESCRIPTION
**Pull request by SleekMutt**
resolves #910 

## Code reviewers

- [ ] @loqimean 

## Summary of issue

Create menstrual hygiene calculator to display 4 calculations: 
1. how many disposable menstrual products have been used since the onset of menstruation until the current age
2. how much money has already been spent on disposable menstrual products from the time of menstruation to the current age
3. how much more menstrual products can be used from the current age to the expected age of menopause
4. how much money can still be spent on menstrual products from the current age to the expected age of menopause

## Summary of change

-Created api controller and action to the existing controller
-Created service to calculate mentioned indicators 
-Created validations to filter invalid values
-Created stimulus controller to the calculator view
-Created view with the calculator 
-Created tests to test calculations

https://github.com/user-attachments/assets/ab50c980-2d3e-405e-b38a-09a96d38f7ca

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [x]  PR is reviewed manually again (to make sure you have 100% ready code)
- [x]  All reviewers agreed to merge the PR
- [x]  I've checked new feature as logged in and logged out user if needed
- [x]  PR meets all conventions

### CODE REVIEW

File: app/controllers/calculators_controller.rb [hunk 1]
```diff
@@ -33,6 +33,15 @@ def calculator
     end
   end
 
+  def mhc_calculator
+    @diaper_categories   = Category.ordered_by_diapers_periods_price
+    @preferable_category = Category.preferable.first
+    add_breadcrumb t("breadcrumbs.home"), root_path
+    add_breadcrumb t(".mhc_calculator.calculator_name")
+
+    render "calculators/mhc_calculator"
```
Comment (loqimean) [inline app/controllers/calculators_controller.rb:42]: а це навіщо? рельса і так знає, що тре зарендерити)

File: app/views/layouts/_mhc_description_block.html.slim [hunk 2]
```diff
@@ -0,0 +1,12 @@
+span
+  p data-type="description_block_html"
+  - if I18n.locale == :en
+    = it ".mhc_description_block_html",
+            information_link: It.link("https://docs.google.com/document/d/1xu6cch-Ner9OEpU4_TV5suIUsuVcWpgiUuX6QrOgaL4/edit",
```
Comment (loqimean) [inline app/views/layouts/_mhc_description_block.html.slim:5]: а що це за документ?

File: app/views/layouts/_mhc_description_button.html.slim [hunk 3]
```diff
@@ -0,0 +1,6 @@
+- if I18n.locale == :en
+  = it "layouts.description_button.use_less",
```
Comment (loqimean) [inline app/views/layouts/_mhc_description_button.html.slim:2]: що це за `it`?)

**Review by loqimean — CHANGES_REQUESTED** [review]

File: app/views/layouts/_mhc_description_block.html.slim (hunk 2 above)
Comment (SleekMutt) [reply app/views/layouts/_mhc_description_block.html.slim:5]: https://docs.google.com/document/d/1q37RP0wfgcHJJxWo49dYbMhc1sPvUVqe/edit

В кінці написано, що треба додати до тексту посилання на цей документ

File: app/views/layouts/_mhc_description_button.html.slim (hunk 3 above)
Comment (SleekMutt) [reply app/views/layouts/_mhc_description_button.html.slim:2]: локалізація

**Review by loqimean — APPROVED** [review]

**Review by loqimean — APPROVED** [review]

File: app/validators/mhc_calculator_validator.rb
Comment (loqimean) [inline app/validators/mhc_calculator_validator.rb:1]: стоп, а де тести на валіадтор?

Comment (loqimean) [reply app/validators/mhc_calculator_validator.rb:1]: весь рубішний код має бути покритий тестами

**Review by loqimean — CHANGES_REQUESTED** [review]

File: spec/validators/mhc_calculator_validator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        validator = described_class.new(valid_params)
+
+        expect(validator.valid?).to be true
+        expect(validator.errors).to be_empty
+      end
+    end
+
+    context "when some parameter is missing" do
+      it "returns false" do
+        invalid_params = valid_params.except(:user_age)
```
Comment (loqimean) [inline spec/validators/mhc_calculator_validator_spec.rb:29]: НІКОЛИ так не пиши, у нас для цього є `let`, я ж і не раз вже казав

File: spec/validators/mhc_calculator_validator_spec.rb
Comment (loqimean) [inline spec/validators/mhc_calculator_validator_spec.rb:1]: не всі методи покриті тестами

**Review by loqimean — CHANGES_REQUESTED** [review]

File: spec/validators/mhc_calculator_validator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        expect(validator.valid?).to be false
+        expect(validator.errors).to include(:user_age)
+        expect(validator.errors[:user_age]).to eq(I18n.t("#{local_error_prefix}.presence_error_msg", field: I18n.t("#{local_field_prefix}.user_age")))
+      end
+    end
+  end
+
+  describe "individual validations" do
+    shared_examples "a presence validation" do |attribute|
```
Comment (loqimean) [inline spec/validators/mhc_calculator_validator_spec.rb:43]: шерид екземпли повинні бути в папці support

File: spec/validators/mhc_calculator_validator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        expect(validator.valid?).to be true
+        expect(validator.errors).to be_empty
+      end
+    end
+
+    context "when some parameter is missing" do
+      let(:invalid_params) { valid_params.except(:user_age, :pads_per_cycle) }
+
+      it "returns false" do
+        validator = described_class.new(invalid_params)
```
Comment (loqimean) [inline spec/validators/mhc_calculator_validator_spec.rb:33]: а це так і не пофіксав

**Review by loqimean — CHANGES_REQUESTED** [review]

File: spec/support/shared/presence_validation_example.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+shared_examples "presence validation" do |attribute|
+  it "#{attribute} is valid when present" do
+    expect(validator.send(:"validate_#{attribute}")).to be true
+    expect(validator.errors[attribute]).to be_nil
+  end
+
+  it "#{attribute} is invalid when blank" do
+    validator.params[attribute] = nil
```
Comment (loqimean) [inline spec/support/shared/presence_validation_example.rb:8]: для цього використовуй before хук

**Review by loqimean — CHANGES_REQUESTED** [review]

**Review by loqimean — APPROVED** [review]

**Pull request merged**